feat: use `/proc/self/cgroup` and `/proc/self/mountinfo` to retrieve the machine-id on Docker environments

mathnogueira opened the pull request:
~At first I was going to use the `/proc/self/cgroup`, however, recent versions of Docker don't keep the container id there anymore and it would result in an empty string as well. The only alternative I found is using `/etc/hostname` instead.~

I also had to change the `readFile` helper function because on an Ubuntu container, the file `/etc/machine-id` exists, but it's empty. So I added a new condition that the file content must not be empty in order to be valid.

Edit: I changed the implementation as we talked and now it looks for `/proc/self/cgroup` and `/proc/self/mountinfo` , just like https://github.com/keygen-sh/py-machineid does.

**Alpine**
![image](https://github.com/user-attachments/assets/0c561b86-6dd8-4cf7-ae46-c7ccd1f841cc)



**Ubuntu**
![image](https://github.com/user-attachments/assets/67e54dd9-0a36-4f13-b344-4502fe915a8f)

mathnogueira committed:
feat: use hostname as a last fallback for Docker

mathnogueira committed:
rephrase comment

mathnogueira committed:
fix: make sure file is not empty

ezekg commented:
>At first I was going to use the `/proc/self/cgroup`, however, recent versions of Docker don't keep the container id there anymore and it would result in an empty string as well.

Are you sure? I haven't seen any reports of this at https://github.com/keygen-sh/py-machineid. It checks both [`/proc/self/cgroup`](https://github.com/keygen-sh/py-machineid/blob/master/machineid/__init__.py#L97) and [`/proc/self/mountinfo`](https://github.com/keygen-sh/py-machineid/blob/master/machineid/__init__.py#L101C29-L101C49), though. Using both of those are going to be a lot more accurate than relying on the hostname, which I don't really think is a good idea because it has no uniqueness guarantees.

mathnogueira commented:
`/proc/self/cgroup` contains `0::/` on Docker. I think `py-machineid` is checking it for compatibility of older Docker versions. Just to be safe, I'll use it as well and if there's nothing there, I will rely on the `mountinfo`.  Thanks for pointing that out. I wasn't happy about using the hostname either.

I tested it out and `/proc/self/mountinfo` is available in both `ubuntu` and `alpine`, so I expect this to be true in the other base images as well.

![image](https://github.com/user-attachments/assets/dd22fb24-732c-4e3e-99d5-e5201951f5cb)

ezekg commented:
Happy to accept a PR that implements the above akin to py-machineid! I'm sure others would love Docker support too.

mathnogueira committed:
fix: use `/proc/self/cgroup` and `/proc/self/mountinfo` to get id

mathnogueira commented:
@ezekg I changed the implementation and now it looks for both `cgroup` and `mountinfo` files just like `py-machineid` does. I also updated the title/description of the PR to reflect the changes. You can see the machine-id generated in both `ubuntu` and `alpine` images in the description of the PR.

ezekg commented:
Tagged v1.1.0.
